src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,246 @@
+      int64_t ret = read[RGW_QOS_BW] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+    int64_t write_bytes()
+    {
+      int64_t ret = write[RGW_QOS_BW] / fixed_float_rgw_ratelimit;
+      return ret;
+    }
+  };
+  const size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::condition_variable *cv;
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
+    if((strcmp(method,"GET") == 0) || (strcmp(method,"HEAD") == 0))
+      return true;
+    return false;
+  }
+
+    // find or create an entry, and return its iterator
+  auto find_or_create(const std::string& key) {
+    std::shared_lock rlock(insert_lock);
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.find(key);
+    rlock.unlock();
+    if(ret == qos_entries.end())
+    {
+      std::unique_lock wlock(insert_lock);
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+    }
+    return ret;
+  }
+
+  bool increase_entry_read(hash_map::iterator& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it->second.read_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it->second.read_bytes() < 0 && bw_limit > 0))
+      return true;
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it->second.read[RGW_QOS_OPS] -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+  bool minimum_time_reached(const ceph::timespan& curr_timestamp, const hash_map::iterator& it)
+  {
+    using namespace std::chrono;
+    constexpr auto min_duration = duration_cast<ceph::timespan>(seconds(60)) / fixed_float_rgw_ratelimit;
+    const auto delta = curr_timestamp - it->second.ts;
+    if (delta < min_duration)
+      return false;
+    return true;
+  }
+  bool increase_tokens(hash_map::iterator& it, const ceph::timespan& curr_timestamp,
+                       const RGWQoSInfo& tenant)
+  {
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
+    if(curr_timestamp > it->second.ts && !it->second.first_run && minimum_time_reached(curr_timestamp, it))
+    {
+      int64_t time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(curr_timestamp - it->second.ts).count() / 60.0 / std::milli::den * fixed_float; // / 60 to make it work with 1 min token bucket
+      it->second.ts = curr_timestamp;
+      int64_t write_ops = tenant.max_write_ops * time_in_ms;
+      int64_t write_bw = tenant.max_write_bytes * time_in_ms;
+      int64_t read_ops = tenant.max_read_ops * time_in_ms;
+      int64_t read_bw = tenant.max_read_bytes * time_in_ms;
```

Comment:
Yes, I have tested and it is working fine with 1(probably because it's combined with higher minimum_time_reached).
Before this change for lower values after 61 seconds of running there were not enough tokens and the tenant still got only 1 request (it took 70~ seconds to get 2 requests)